### PR TITLE
Rails 5.2

### PR DESCRIPTION
**Pull request by ik5**
Based on @mmoll [work](https://github.com/theforeman/foreman/pull/5428), work in progress for the skeleton.

Please wait until the WIP will be removed for review, it is going to change in some places before it can be reviewed properly.

### CODE REVIEW

**Comment by tbrisker** [comment]
@mmoll - are you planning to continue with the 5.2 migration? if so, please coordinate between the two of you so you don't duplicate the work.

**Comment by mmoll** [comment]
If I have something, I can also PR it into this PR :+1:

**Comment by tbrisker** [comment]
@mmoll @ik5 - if there are parts that are backwards compatible with 5.1 I think it would be best to split them into separate PRs that we can merge prior to the actual upgrade, so that this pr will only contain the breaking changes.

**Comment by ik5** [comment]
@tbrisker
The current work looks like that it is for non compatible features, but
there are few things that still does not pass the tests, that I'm on them
at the moment.

File: test/models/concerns/key_pair_compute_resource_test.rb
Comment (ik5) [inline test/models/concerns/key_pair_compute_resource_test.rb]: @dLobatog this test fails, after rails 5.2 [fixed the `destroy!`](https://github.com/rails/rails/commit/b0fc04aa3af338d5a90608bf37248668d59fc881) method, and I'm unable to make it work.

Can you please take a look at your old code and maybe notice something that I'm missing?

Comment (dLobatog) [reply test/models/concerns/key_pair_compute_resource_test.rb]: I think when I did this test, I did way too much mocking. Thankfully `Fog.mock` works very well these days, so I changed it to simply use `Fog.mock` and `unmock`. This ensures that all the methods calling fog are using the right fog objects in tests, but the API calls are faked (they return some fake JSON that is parsed by fog).

Also it looks cleaner and harder to break the previous test :smile: 
```
diff --git a/test/models/concerns/key_pair_compute_resource_test.rb b/test/models/concerns/key_pair_compute_resource_test.rb
index a51343932..e8ae6b691 100644
--- a/test/models/concerns/key_pair_compute_resource_test.rb
+++ b/test/models/concerns/key_pair_compute_resource_test.rb
@@ -5,7 +5,10 @@ class DummyComputeResource < ComputeResource
   include Mocha::API

   def client
-    @client ||= mock('client')
+    @client ||= ::Fog::Compute.new(
+      :provider => :aws,
+      :aws_access_key_id => 'foo',
+      :aws_secret_access_key => 'bar')
   end
 end

@@ -36,15 +39,11 @@ class KeyPairComputeResourceTest < ActiveSupport::TestCase
   end

   test 'should remove the key pair on compute resource deletion' do
+    Fog.mock!
     key_pair = FactoryBot.create(:key_pair)
     cr = DummyComputeResource.create(key_pair: key_pair, name: Foreman.uuid, provider: 'EC2')
     cr.update_attribute(:key_pair, key_pair)
-    mock_key_pairs = mock('mock_key_pairs')
-    fog_key_pair = mock('fog_key_pair')
-    cr.send(:client).expects(:key_pairs).returns(mock_key_pairs)
-    mock_key_pairs.expects(:get).with(key_pair.name).returns(fog_key_pair)
-    key_pair.expects(:destroy).once.returns(true)
-    fog_key_pair.expects(:destroy).once.returns(true)
+    Fog.unmock!
     assert cr.destroy!
   end
 end
```

Comment (ik5) [reply test/models/concerns/key_pair_compute_resource_test.rb]: cool it works thank you very much!

**Comment by ik5** [comment]
[test foreman]

**Comment by iNecas** [comment]
@ik5 could we get back to the point where all the changes were there as separate commits + the new one based on https://github.com/theforeman/foreman/pull/5836#discussion_r206854123.

It would help the review.

**Comment by ik5** [comment]
@iNecas, nop the history re-rewritten :( 
Tried locally, the entire original rebase got lost

**Comment by ik5** [comment]
@iNecas here you go

**Comment by mmoll** [comment]
@dLobatog @ik5 if the test fix is compatible with Rails 5.1, please open a separate PR against develop for it, the redmine issue for it is https://projects.theforeman.org/issues/22938

**Comment by iNecas** [comment]
The integration tests fail due to Dynflow initialization during `db:create`. This should help:

```patch
diff --git a/config/application.rb b/config/application.rb
index a168b3291..ce58af407 100644
--- a/config/application.rb
+++ b/config/application.rb
@@ -278,7 +278,7 @@ module Foreman
     end
 
     config.after_initialize do
-      init_dynflow unless Foreman.in_rake?('db:drop')
+      init_dynflow unless Foreman.in_rake?('db:create') || Foreman.in_rake?('db:drop')
       setup_auditing
     end

```

**Comment by mmoll** [comment]
opened GH-5912 and GH-5913

**Comment by iNecas** [comment]
The katello tests fail to initialize. Seems to be also related to some changes with `rake db:create`, perhaps we need to check on `Foreman.in_rake('db:create')` in https://github.com/theforeman/foreman/blob/f715a7c98457c6f43a1f803a75ba4540372d8544/app/registries/foreman/plugin.rb#L323

Otherwise if fails when calling `rake db:create` when some plugin registered.

```
07:06:00 ActiveRecord::NoDatabaseError: FATAL:  database "test_develop_pr_katello-0-dev" does not exist
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/postgresql_adapter.rb:688:in `rescue in connect'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/postgresql_adapter.rb:683:in `connect'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/postgresql_adapter.rb:215:in `initialize'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/postgresql_adapter.rb:40:in `new'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/postgresql_adapter.rb:40:in `postgresql_connection'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/abstract/connection_pool.rb:809:in `new_connection'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/abstract/connection_pool.rb:853:in `checkout_new_connection'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/abstract/connection_pool.rb:832:in `try_to_checkout_new_connection'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/abstract/connection_pool.rb:793:in `acquire_connection'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/abstract/connection_pool.rb:521:in `checkout'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/abstract/connection_pool.rb:380:in `connection'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/abstract/connection_pool.rb:1008:in `retrieve_connection'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_handling.rb:118:in `retrieve_connection'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_handling.rb:90:in `connection'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/schema_migration.rb:22:in `table_exists?'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/schema_migration.rb:26:in `create_table'
07:06:00 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/migration.rb:1187:in `initialize'
07:06:00 /var/lib/workspace/workspace/test_develop_pr_katello/database/postgresql/ruby/2.4/slave/fast/app/registries/foreman/plugin.rb:325:in `new'
07:06:00 /var/lib/workspace/workspace/test_develop_pr_katello/database/postgresql/ruby/2.4/slave/fast/app/registries/foreman/plugin.rb:325:in `pending_migrations'
07:06:00 /var/lib/workspace/workspace/test_develop_pr_katello/database/postgresql/ruby/2.4/slave/fast/app/registries/foreman/plugin.rb:269:in `permission'
```

**Comment by mmoll** [comment]
@ik5 I rebased onto current develop and removed one commit that's not needed anymore.

**Comment by ik5** [comment]
@mmoll i'm not sure that I'm able to follow on what you are doing.
Some of the commits you have made, I thought I already made them for example.

Nice to know about the deprecation yaml file :)

**Comment by mmoll** [comment]
0fc09aa is not needed anymore, what do you mean with "Some of the commits you have made, I thought I already made them"?

**Comment by iNecas** [comment]
@ik5 I guess what's @mmoll is asking for is rebasing this PR against develop branch: that should make some commits, that were merged to develop in the meantime, disappear from this PR.

Also, it seems, that the one difference Rails 5.2 has is it initializes the rails environment during the `db:create`, therefore, we should probably search for presence of `if Foreman.in_rake?('db:migrate')` and add `Foreman.in_rake?('db:create')` as well, to get over the setup failures.

**Comment by mmoll** [comment]
I did rebase and force-push, sorry if that wasn't clear.

**Comment by ik5** [comment]
@mmoll, @iNecas, okay, did some rebase on develop, and also push force, (also pull rebase first), I hope i haven't missed anything

**Comment by iNecas** [comment]
Here is the PR for addressing the db:create issue with plugins  https://github.com/theforeman/foreman/pull/5947

File: test/models/role_test.rb
Comment (iNecas) [inline test/models/role_test.rb]: I guess we need a new expectation here: otherwise the test just doesn't make sense…

**Comment by mmoll** [comment]
@iNecas seems like it's also needed for `db:drop`...

**Comment by iNecas** [comment]
I can do that: on the other hand, that's not the primary cause for the tests to not continue. The fatal errors happens during db:create, where bastion has some additional code that needs update as well to use the new method: 

```
10:58:58 ActiveRecord::NoDatabaseError: FATAL:  database "test_develop_pr_katello-0-dev" does not exist
10:58:58 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/postgresql_adapter.rb:688:in `rescue in connect'
10:58:58 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/postgresql_adapter.rb:683:in `connect'
10:58:58 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/postgresql_adapter.rb:215:in `initialize'
10:58:58 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/postgresql_adapter.rb:40:in `new'
10:58:58 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/postgresql_adapter.rb:40:in `postgresql_connection'
10:58:58 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/abstract/connection_pool.rb:809:in `new_connection'
10:58:58 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/abstract/connection_pool.rb:853:in `checkout_new_connection'
10:58:58 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/abstract/connection_pool.rb:832:in `try_to_checkout_new_connection'
10:58:58 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/abstract/connection_pool.rb:793:in `acquire_connection'
10:58:58 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/abstract/connection_pool.rb:521:in `checkout'
10:58:58 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/abstract/connection_pool.rb:380:in `connection'
10:58:58 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_adapters/abstract/connection_pool.rb:1008:in `retrieve_connection'
10:58:58 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_handling.rb:118:in `retrieve_connection'
10:58:58 /usr/local/rvm/gems/ruby-2.4.3@test_develop_pr_katello-0/gems/activerecord-5.2.0/lib/active_record/connection_handling.rb:90:in `connection'
10:58:58 /var/lib/workspace/workspace/test_develop_pr_katello@2/database/postgresql/ruby/2.4/slave/fast/plugin/engines/bastion_katello/lib/bastion_katello/engine.rb:13:in `block in <class:Engine>'
```
See https://github.com/Katello/katello/blob/333f03036e4f5659e9b9e4dc92b9bea8ac7a0324/engines/bastion_katello/lib/bastion_katello/engine.rb#L13

**Comment by mmoll** [comment]
[test katello]

**Comment by ik5** [comment]
There are weird issues with the setup of a testing environment (didn't test the development part).

Sometimes even Katello migration (adding Katello after db:create) fails to do alter table if drop is needed first, and sometimes t just works, no basic rule that I could figure out.

Also, the seeding does not add the content to the database at all.

**Comment by ik5** [comment]
following additional testing it does execute seeds, but act like they are not existed, still investigating it

**Comment by iNecas** [comment]
The `db:seed` worked for me.

**Comment by iNecas** [comment]
The Katello part of the fix for db:create is available here https://github.com/Katello/katello/pull/7639

**Comment by iNecas** [comment]
@ik5 could you rebase again? The jenkins is currently failling due to issue, that was resolved in the develop branch in the mean time (https://github.com/theforeman/foreman/pull/5962)

**Comment by ik5** [comment]
@iNecas rebased and pushed it an hour ago +-

File: Gemfile
Comment (iNecas) [reply Gemfile]: We should probably disable the rubocop warning here.

Comment (mmoll) [reply Gemfile]: I guess this time we'll go from 5.1 to 5.2 straight, so we can also just replace 5.1 with 5.2

Comment (iNecas) [reply Gemfile]: Ok, in that case we can just go ahead and do that.

**Comment by mmoll** [comment]
[test katello]

**Comment by mmoll** [comment]
[test upgrade]

File: Gemfile
Comment (tbrisker) [reply Gemfile]: I agree - go straight to 5.2.1 IMHO.

File: config/as_deprecation_whitelist.yaml
Comment (tbrisker) [inline config/as_deprecation_whitelist.yaml]: Where is this warning from? sound pretty scary to me

Comment (ik5) [reply config/as_deprecation_whitelist.yaml]: Rails 5.2 changed that, so we are adding it here

Comment (mmoll) [reply config/as_deprecation_whitelist.yaml]: @tbrisker See e.g. https://travis-ci.org/mmoll/foreman/jobs/412885124 at the bottom for the actual output. These should ideally get fixed after 5.2 is merged.

Comment (tbrisker) [reply config/as_deprecation_whitelist.yaml]: Thanks, can you open an issue to track that please?

Comment (mmoll) [reply config/as_deprecation_whitelist.yaml]: https://projects.theforeman.org/issues/23300 is more or less this.

File: Gemfile
Comment (mmoll) [inline Gemfile:9]: this also needs adjustments elsewhere

**Comment by ehelms** [comment]
FYI: Foreman RPM built against Ruby 2.5 and Rails 5.2 with this patch applied locally.

File: test/models/role_test.rb
Comment (ik5) [reply test/models/role_test.rb]: done?

Comment (iNecas) [reply test/models/role_test.rb]: Works for me

**Comment by iNecas** [comment]
There are 3 tests failing right now for this branch (https://ci.theforeman.org/job/test_develop_pr_core/19097/database=mysql,ruby=2.4,slave=fast/), all failing with message 'ActiveRecord::RecordInvalid: Validation failed: Value is not a number'.

@ik5 could you rebase this to latest develop, as the develop branch seems to be green. If the issue persists,  this could be something related to Rails 5.2.1 or we've not seen this in 5.2.0

**Comment by mmoll** [comment]
Yeah, it's a new error with 5.2.1, regarding numerical values and how defaults are stored with the settings currently.

File: Gemfile
Comment (tbrisker) [reply Gemfile:9]: Specifically in `config/settings.yaml.example`

**Comment by tbrisker** [comment]
The issue came from this change in rails: https://github.com/rails/rails/commit/c15d8b3a21bf1ae3ca9f09ca9cad8cc3828aa9be#diff-58f66170fe2bce77e8a8b3c69b77a0d6

**Comment by tbrisker** [comment]
To fix the issue, add the following method to `app/models/setting.rb`:
```ruby
  def read_attribute_before_type_cast(attr_name)
    return value if attr_name == :value
    super(attr_name)
  end
```

**Comment by tbrisker** [comment]
and tbh we do some crazy voodoo in `setting.rb` that might be a good idea to refactor... (most likely some stuff left from ruby 3 days...)

**Comment by mmoll** [comment]
I opened GH-6004 and GH-6006. Once these are in, this PR can get rebased and everything is IMHO ready. :heart:

@ik5 I force-pushed into this branch, please be sure to update your local copy.

**Comment by tbrisker** [comment]
[test foreman] most likely transient but better make sure

**Comment by mmoll** [comment]
I *guess* the test failures are the same transient errors as in current develop, so it would be better first stabilize the test with Rails 5.1.

**Review by tbrisker — CHANGES_REQUESTED** [review]
Different transient failures this time, so :+1: from me. Setting to Request Changes just so merging is blocked until we coordinate the merge (since this will break all plugins that haven't been fixed yet)

**Comment by tbrisker** [comment]
[test katello] will hopefully pass now that katello side has been merged.

**Comment by tbrisker** [comment]
[test katello]

**Comment by iNecas** [comment]
@ik5 the katello tests are failing with "ERROR:  cached plan must not change result type", I remember you had the fix with `reset_column_information` in Katello's PR originally: not sure we it got out: perhaps some issue during the rebase?

**Comment by ik5** [comment]
@iNecas looking on this

**Comment by ik5** [comment]
Here is my fix: https://github.com/Katello/katello/pull/7663

**Comment by tbrisker** [comment]
[test katello]

**Comment by mmoll** [comment]
katello is :green_heart: now

**Comment by tbrisker** [comment]
Thanks @ik5 @mmoll @iNecas!

**Review by tbrisker — APPROVED** [review]

**Comment by boaz0** [comment]
congrats

**Comment by dLobatog** [comment]
Great job @ik5 @mmoll @iNecas !

**Pull request merged**
